### ISSUE TITLE
[Bug] addAccount failed when calling GetAccountsAsync()

### ISSUE THREAD

**Issue by markbeij** [AzureAD/microsoft-authentication-library-for-dotnet#1856]
**Which Version of MSAL are you using ?**
MSAL 4.14.0

**Platform**
Xamarin Forms mobile app (Android Only)

**What authentication flow has the issue?**
Issues occurs at `GetAccountsAsync()`, before `AcquireTokenSilent(`) or `AcquireTokenInteractive()` are called.

**Is this a new or existing app?**
This app is in production. Currently still on ADAL. We're currently working on upgrade to MSAL.

**Repro**

```csharp
private async Task<string> GetAccessTokenAsyncInternal(params string[] scopes)
{
    var msal = CreateMsalClient();

    var accounts = await msal.GetAccountsAsync();
    AuthenticationResult result;
    try
    {
        result = await msal.AcquireTokenSilent(scopes, accounts.FirstOrDefault())
                    .ExecuteAsync();
        return result.AccessToken;
    }
    catch (MsalServiceException ex)
    {
        if (!(ex is MsalUiRequiredException) && ex.ErrorCode != MsalError.BrokerResponseReturnedError)
            throw;

        result = await msal.AcquireTokenInteractive(scopes)
            .WithUseEmbeddedWebView(true)
            .WithParentActivityOrWindow(currentActivityOrWindowProvider.GetCurrentActivityOrWindow())
            .ExecuteAsync();
        return result.AccessToken;
    }
}

private IPublicClientApplication CreateMsalClient()
{
    var isBrokerAvailable = brokerService.DetectBroker();

    return PublicClientApplicationBuilder
        .Create(clientId)
        .WithBroker(isBrokerAvailable)
        .WithAuthority(authority, true)
        .WithIosKeychainSecurityGroup("xx.yy.zz")
        .WithRedirectUri(redirectUri)
        .WithParentActivityOrWindow(() => currentActivityOrWindowProvider.GetCurrentActivityOrWindow())
        .Build();
}


private const string PACKAGE = "com.azure.authenticator";

public bool DetectBroker()
{
    bool hasBroker;

    try
    {
        PackageManager pm = CrossCurrentActivity.Current.Activity.PackageManager;
        pm.GetPackageInfo(PACKAGE, PackageInfoFlags.Activities);
        hasBroker = true;
    }
    catch (PackageManager.NameNotFoundException)
    {
        hasBroker = false;
    }

    Analytics.TrackEvent("Broker Installed", new Dictionary<string, string>
    {
        {"Broker", PACKAGE},
        {"installed", hasBroker.ToString()}
    });

    return hasBroker;
}
```

**Expected behavior**
A clear and concise description of what you expected to happen (or code).

**Actual behavior**
`IPublicClientApplication.GetAccountsAsync()` throws an exception.
```json
"exception": {
  "type": "Microsoft.Identity.Client.MsalClientException",
  "message": "addAccount failed",
  "stackTrace": "  at Microsoft.Identity.Client.Platforms.Android.AndroidBroker.HandleBrokerOperationError (System.Exception ex) [0x0002b] in <62c5f81aa08b40f6a5628154f09a4f0a>:0 \n  at Microsoft.Identity.Client.Platforms.Android.AndroidBroker.GetAccountsAsync (System.String clientID) [0x000ef] in <62c5f81aa08b40f6a5628154f09a4f0a>:0 \n  at Microsoft.Identity.Client.ClientApplicationBase.GetAccountsAsync () [0x000cf] in <62c5f81aa08b40f6a5628154f09a4f0a>:0 \n  at someapp.Authentication.MsalAuthenticator.GetAccessTokenAsyncInternal (System.String[] scopes) [0x00096] in <7d586e783c0a4d1eb572d612ea9ef867>:0 \n  at Polly.Retry.AsyncRetryEngine.ImplementationAsync[TResult] (System.Func`3[T1,T2,TResult] action, Polly.Context context, System.Threading.CancellationToken cancellationToken, Polly.ExceptionPredicates shouldRetryExceptionPredicates, Polly.ResultPredicates`1[TResult] shouldRetryResultPredicates, System.Func`5[T1,T2,T3,T4,TResult] onRetryAsync, System.Int32 permittedRetryCount, System.Collections.Generic.IEnumerable`1[T] sleepDurationsEnumerable, System.Func`4[T1,T2,T3,TResult] sleepDurationProvider, System.Boolean continueOnCapturedContext) [0x000c3] in <12c445b8cfb1467e873f6690adeddb4b>:0 ",
  "innerExceptions": [
    {
      "type": "Android.Accounts.AuthenticatorException",
      "message": "addAccount failed",
      "stackTrace": "  at Java.Interop.JniEnvironment+InstanceMethods.CallObjectMethod (Java.Interop.JniObjectReference instance, Java.Interop.JniMethodInfo method) [0x00068] in <26521a5118b44c858c385715922b9d5d>:0 \n  at Android.Runtime.JNIEnv.CallObjectMethod (System.IntPtr jobject, System.IntPtr jmethod) [0x0000e] in <4ccdb3137d974856b786e1aeebbfbab6>:0 \n  at Android.Accounts.IAccountManagerFutureInvoker.get_Result () [0x00033] in <4ccdb3137d974856b786e1aeebbfbab6>:0 \n  at Microsoft.Identity.Client.Platforms.Android.AndroidBrokerHelper.GetBrokerAccounts (System.Collections.Generic.IDictionary`2[TKey,TValue] brokerPayload, Android.App.Activity callerActivity) [0x0003c] in <62c5f81aa08b40f6a5628154f09a4f0a>:0 \n  at Microsoft.Identity.Client.Platforms.Android.AndroidBrokerHelper.GetBrokerAccountsInAccountManager (System.Collections.Generic.IDictionary`2[TKey,TValue] brokerPayload) [0x00000] in <62c5f81aa08b40f6a5628154f09a4f0a>:0 \n  at Microsoft.Identity.Client.Platforms.Android.AndroidBroker.GetAccountsAsync (System.String clientID) [0x000c9] in <62c5f81aa08b40f6a5628154f09a4f0a>:0 \n  --- End of managed Android.Accounts.AuthenticatorException stack trace ---\nandroid.accounts.AuthenticatorException: addAccount failed\n\tat android.accounts.AccountManager.convertErrorToException(AccountManager.java:2511)\n\tat android.accounts.AccountManager.-wrap0(Unknown Source:0)\n\tat android.accounts.AccountManager$AmsTask$Response.onError(AccountManager.java:2347)\n\tat android.accounts.IAccountManagerResponse$Stub.onTransact(IAccountManagerResponse.java:69)\n\tat android.os.Binder.execTransact(Binder.java:682)\n",
      "wrapperSdkName": "appcenter.xamarin"
    }
  ],
  "wrapperSdkName": "appcenter.xamarin"
}
```

**Additional context/ Logs / Screenshots**
Tested this on several devices, but only one device has this issue.
It's a Samsung S7, Android 8.0.0. This device has Intune Company Portal installed as well as MS Authenticator.
When we remove MS Authenticator from this device this issue does not occur.

The stack trace does not give me the impression that any broker is used, although I'm sure `withBroker` is `true`.
```
  at Java.Interop.JniEnvironment+InstanceMethods.CallObjectMethod (Java.Interop.JniObjectReference instance, Java.Interop.JniMethodInfo method) [0x00068] in <b294f53475ed47848b5b242d7de1c041>:0 
    at Android.Runtime.JNIEnv.CallObjectMethod (System.IntPtr jobject, System.IntPtr jmethod) [0x0000e] in <137765e30915460db72acf0bb1f3cb78>:0 
      at Android.Accounts.IAccountManagerFutureInvoker.get_Result () [0x00033] in <137765e30915460db72acf0bb1f3cb78>:0 
        at Microsoft.Identity.Client.Platforms.Android.AndroidBrokerHelper.GetBrokerAccounts (System.Collections.Generic.IDictionary`2[TKey,TValue] brokerPayload, Android.App.Activity callerActivity) [0x0003c] in <62c5f81aa08b40f6a5628154f09a4f0a>:0 
          at Microsoft.Identity.Client.Platforms.Android.AndroidBrokerHelper.GetBrokerAccountsInAccountManager (System.Collections.Generic.IDictionary`2[TKey,TValue] brokerPayload) [0x00000] in <62c5f81aa08b40f6a5628154f09a4f0a>:0 
            at Microsoft.Identity.Client.Platforms.Android.AndroidBroker.GetAccountsAsync (System.String clientID) [0x000c9] in <62c5f81aa08b40f6a5628154f09a4f0a>:0 
          --- End of managed Java.Lang.Exception stack trace ---
          android.accounts.AuthenticatorException: addAccount failed
          \tat android.accounts.AccountManager.convertErrorToException(AccountManager.java:2511)
          \tat android.accounts.AccountManager.-wrap0(Unknown Source:0)
          \tat android.accounts.AccountManager$AmsTask$Response.onError(AccountManager.java:2347)
          \tat android.accounts.IAccountManagerResponse$Stub.onTransact(IAccountManagerResponse.java:69)
          \tat android.os.Binder.execTransact(Binder.java:682)
```

**Comment by jmprieur** [comment]
@markbeij are you using the latest version of MS communicator?

**Comment by markbeij** [comment]
I believe you mean MS Authenticator. Yes we use the latest.
I don't believe MS Authenticator could be the issue though, as the exceptions comes from MSAL calling `android.accounts.AccountManager`

**Comment by bgavrilMS** [comment]
MSAL uses the Android AccountManager to find out what accounts are present in the Authentication app. [It does this by calling](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/04c30020aff2ac9859e024b7ec000c9c12a5667c/src/client/Microsoft.Identity.Client/Platforms/Android/AndroidBrokerHelper.cs#L84) `AccountManager.AddAccount`. 

If for some reason the OS is blocking calls to AccountManager (e.g. Android for Work), MSAL will fail. Can you use the AccountManager in your app to add an account just to test that the call is working?

Note that ADAL also uses the AccountManager, but it calls `GetAccounts` on it, which requires a ton of permissions (account picker etc.). `AddAccount` does not require permissions. But still, AccountManager is used.

**Comment by markbeij** [comment]
I'll check if we can try this

**Comment by markbeij** [comment]
To see if `AccountManager.AddAccount` fails when running from my own code, I added some logic in the `MainActivity`

```csharp
protected override void OnCreate(Bundle savedInstanceState)
{
    base.OnCreate(savedInstanceState);
    ...
    AndroidConfiguration configuration = new AndroidConfiguration();
    _app = new App(new StandardKernel(configuration));
    LoadApplication(_app);
    ...
    Microsoft.AppCenter.AppCenter.Start("ios=f11e9024-b55a-4c3a-8fcd-510f3b6885c0;" +
                    "android=285c4382-81e2-4a53-9044-872220138890;",
                    typeof(Microsoft.AppCenter.Analytics.Analytics), typeof(Microsoft.AppCenter.Crashes.Crashes));

    ShowAccountsAsync().Wait();
    1.ToString();
}

private async System.Threading.Tasks.Task ShowAccountsAsync()
{
    try
    {
        await System.Threading.Tasks.Task.Run(() =>
        {
            var manager = Android.Accounts.AccountManager.Get(this);
            Bundle addAccountOptions = new Bundle();
            addAccountOptions.PutString("com.microsoft.broker_accountmanager_operation_key", "GET_INTENT_FOR_INTERACTIVE_REQUEST");
            var result = manager.AddAccount("com.microsoft.workaccount", "adal.authtoken.type", null, addAccountOptions, null, null, null);
            var x = result.GetResult(5 * 60, Java.Util.Concurrent.TimeUnit.Seconds);
        }).ConfigureAwait(false);
        Microsoft.AppCenter.Analytics.Analytics.TrackEvent("ShowAccountsAsync_Succeeded");
    }
    catch (Exception ex)
    {
        Microsoft.AppCenter.Crashes.Crashes.TrackError(ex);
    }
}
```

Surprisingly everything goes well and App Center analytics is called to track the event.

**Comment by markbeij** [comment]
Worth to mention that the app permissions are all enabled (forced by policy).
Other users see this as well but don't have problems using MSAL with broker.

![Screenshot_20200611-092727_Package installer](https://user-images.githubusercontent.com/20302465/84357874-9d59be00-abc6-11ea-97fe-fb0ceee7de8f.jpg)

**Comment by bgavrilMS** [comment]
I am having problems reproducing this issue. Please note that if you set `WIthBroker(true)` but a broker is not installed, then MSAL will use a browser instead. The stack trace which you show is definitely broker related, as MSAL does not call Android's `AccountManager` except for broker scenarios.

**Comment by henrik-me** [comment]
@markbeij we are not able to reproduce this issue. Closing, however if you have more information and still need our assistance pls. add more details and re-open.

**Comment by WilkoThomassen** [comment]
I'm experiencing this issue. User has a Galaxy S7, Android 8.0.0.0

Log from AppCenter:

{
  "length": 0,
  "offset": 0,
  "id": "7eef2702-44c6-4f14-8ecd-9e30035835c7",
  "exception": {
    "type": "Microsoft.Identity.Client.MsalClientException",
    "message": "addAccount failed",
    "stackTrace": "  at Microsoft.Identity.Client.Platforms.Android.Broker.AndroidBroker.HandleBrokerOperationError (System.Exception ex) [0x00036] in <03e19d5ee89744e9aa60cf742bfa5fd0>:0 \n  at Microsoft.Identity.Client.Platforms.Android.Broker.AndroidBroker.GetAccountsAsync (System.String clientID, System.String redirectUri) [0x00111] in <03e19d5ee89744e9aa60cf742bfa5fd0>:0 \n  at Microsoft.Identity.Client.ClientApplicationBase.MergeWithBrokerAccountsAsync (System.Collections.Generic.IEnumerable`1[T] accountsFromCache, System.String homeAccountIdFilter) [0x000ed] in <03e19d5ee89744e9aa60cf742bfa5fd0>:0 \n  at Microsoft.Identity.Client.ClientApplicationBase.GetAccountsInternalAsync (Microsoft.Identity.Client.TelemetryCore.Internal.Events.ApiEvent+ApiIds apiId, System.String homeAccountIdFilter) [0x000f2] in <03e19d5ee89744e9aa60cf742bfa5fd0>:0 \n  at Microsoft.Identity.Client.ClientApplicationBase.GetAccountsAsync () [0x00072] in <03e19d5ee89744e9aa60cf742bfa5fd0>:0 \n  at Ns.Ccm.Mercurius.Authentication.MsalAuthenticator.GetAccessTokenAsyncInternal (System.String[] scopes) [0x00088] in <4276d927d8bf45e2acf1f5a84a2b6ed1>:0 \n  at Polly.Retry.AsyncRetryEngine.ImplementationAsync[TResult] (System.Func`3[T1,T2,TResult] action, Polly.Context context, System.Threading.CancellationToken cancellationToken, Polly.ExceptionPredicates shouldRetryExceptionPredicates, Polly.ResultPredicates`1[TResult] shouldRetryResultPredicates, System.Func`5[T1,T2,T3,T4,TResult] onRetryAsync, System.Int32 permittedRetryCount, System.Collections.Generic.IEnumerable`1[T] sleepDurationsEnumerable, System.Func`4[T1,T2,T3,TResult] sleepDurationProvider, System.Boolean continueOnCapturedContext) [0x000c3] in <12c445b8cfb1467e873f6690adeddb4b>:0 ",
    "innerExceptions": [
      {
        "type": "Java.Lang.Exception",
        "message": "addAccount failed",
        "stackTrace": "  at Java.Interop.JniEnvironment+InstanceMethods.CallObjectMethod (Java.Interop.JniObjectReference instance, Java.Interop.JniMethodInfo method) [0x00068] in <4a4ee17d879e437495621c7f1aac77ed>:0 \n  at Android.Runtime.JNIEnv.CallObjectMethod (System.IntPtr jobject, System.IntPtr jmethod) [0x0000e] in <dccb6bc82cf740e88912213390dae18e>:0 \n  at Android.Accounts.IAccountManagerFutureInvoker.get_Result () [0x00033] in <dccb6bc82cf740e88912213390dae18e>:0 \n  at Microsoft.Identity.Client.Platforms.Android.Broker.AndroidBrokerHelper.GetBrokerAccounts (Microsoft.Identity.Client.Platforms.Android.Broker.BrokerRequest brokerRequest, Android.App.Activity callerActivity) [0x0003c] in <03e19d5ee89744e9aa60cf742bfa5fd0>:0 \n  at Microsoft.Identity.Client.Platforms.Android.Broker.AndroidBrokerHelper.GetBrokerAccountsInAccountManager (Microsoft.Identity.Client.Platforms.Android.Broker.BrokerRequest brokerRequest) [0x00000] in <03e19d5ee89744e9aa60cf742bfa5fd0>:0 \n  at Microsoft.Identity.Client.Platforms.Android.Broker.AndroidBroker.GetAccountsAsync (System.String clientID, System.String redirectUri) [0x000eb] in <03e19d5ee89744e9aa60cf742bfa5fd0>:0 \n  --- End of managed Java.Lang.Exception stack trace ---\nandroid.accounts.AuthenticatorException: addAccount failed\n\tat android.accounts.AccountManager.convertErrorToException(AccountManager.java:2511)\n\tat android.accounts.AccountManager.-wrap0(Unknown Source:0)\n\tat android.accounts.AccountManager$AmsTask$Response.onError(AccountManager.java:2347)\n\tat android.accounts.IAccountManagerResponse$Stub.onTransact(IAccountManagerResponse.java:69)\n\tat android.os.Binder.execTransact(Binder.java:682)\n",
        "wrapperSdkName": "appcenter.xamarin"
      }
    ],
    "wrapperSdkName": "appcenter.xamarin"
  },
  "properties": {
    "retryCount": "2",
    "errorCode": "android_broker_operation_failed"
  },
  "appId": "b6db8c99-203e-4ee3-b25e-f6e9fd9a29c6",
  "installId": "f26f7ca4-51c2-4911-a6f9-065e9db68e09",
  "isTestMessage": false,
  "timestamp": "2020-09-30T13:24:30.684Z",
  "sid": "a578558e-9f86-41b9-ae71-a6bc90b0a8f3",
  "device": {
    "sdkName": "appcenter.android",
    "sdkVersion": "3.3.0",
    "wrapperSdkVersion": "3.4.1",
    "wrapperSdkName": "appcenter.xamarin",
    "model": "SM-G930F",
    "oemName": "samsung",
    "osName": "Android",
    "osVersion": "8.0.0",
    "osBuild": "R16NW",
    "osApiLevel": 26,
    "locale": "nl_NL",
    "timeZoneOffset": 120,
    "screenSize": "1080x1920",
    "appVersion": "1.8.12",
    "carrierName": "KPN",
    "carrierCountry": "nl",
    "appBuild": "249862",
    "appNamespace": "ns.ccm.mercurius.droid",
    "wrapperRuntimeVersion": "11.0.2.0"
  }
}

**Comment by markbeij** [comment]
Please reopen as this still occurs. @WilkoThomassen added extra info. Hope this helps.
This occurs in 4.17.1 as well.

**Comment by jmprieur** [comment]
@bgavrilMS @trwalke @neha-bhargava could it be due to power optimization?

**Comment by trwalke** [comment]
I think it may be due to linker settings in the app. Sometimes the linker can link away the types thrown from the xamarin methods resulting in strange looking errors like this. The account manager "add account" operation has 3 exceptions that it throws, which all inherit from "Java.Lang.Exception":

- android.accounts.OperationCanceledException

- android.accounts.AuthenticatorException

- java.io.IOException

It is possible to catch these types explicitly with
```
try {
    var r = result.GetResult (…);
    }
catch (Android.Accounts.OperationCanceledException e) {
    // …
}
```

The other option would be to change the linker settings. 

Looking at the ADAL code I can see that we are already catching the android.accounts.OperationCanceledException so the issue must be coming from the other two

**Comment by bgavrilMS** [comment]
This has been fixed in MSAL 4.27+ as we've added a different method of communication between MSAL and the broker.
